In [69]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score
from sklearn import preprocessing
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor

pd.options.display.float_format = '{:,.2f}'.format

# setup interactive notebook mode
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

from IPython.display import display, HTML

In [70]:
midterm_data = pd.read_excel('Kaggle_Data/demand_pred_midterm.xlsx')
len(midterm_data)
midterm_data.isna().sum().sum()

4825

0

-------------------------------------
3) a) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer:<br>
Rows in dataset: 4825 <br>
NaN Values: 0 <br>

---------------------------------------

In [71]:
midterm_data['logLikes'] = np.log(midterm_data['Likes'])
midterm_data['logPopulation'] = np.log(midterm_data['Population'])
midterm_data['HoursMedianProduct'] = midterm_data['Hours open'] * midterm_data['Median household']
midterm_data.drop(['Likes', 'Population', 'Hours open', 'Median household'], axis=1, inplace=True)

midterm_data['logLikes'].iloc[0]
midterm_data['logPopulation'].iloc[0]
midterm_data['HoursMedianProduct'].iloc[0]

6.22455842927536

8.420902531097951

40

-------------------------------------
3) b) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 <br>

Answer:<br>
'Likes' Logarithm for first row: 6.22455842927536 <br>
'Population' Logarithm for first row: 8.420902531097951 <br>
Product of 'Hours Open' and 'Median Household': 40 <br>


---------------------------------------

In [72]:
midterm_data = pd.get_dummies(midterm_data, columns=['Airport', 'Season'], drop_first=True)
len(midterm_data.columns)

14

-------------------------------------
3) c) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 <br>

Answer:<br>
Number of columns: 14

---------------------------------------

In [73]:
X = midterm_data.drop('Demand', axis=1)
y = midterm_data['Demand']  

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=9729)
X_test.shape[0]

965

-------------------------------------
3) d) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer:<br>
Rows in test data: 965

---------------------------------------

In [74]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model = LinearRegression()
model.fit(X_train_scaled, y_train)
y_train_pred = model.predict(X_train_scaled)
r2_score(y_train, y_train_pred)
model.intercept_

LinearRegression()

0.556471669081702

143.51450777202072

-------------------------------------
3) e) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 <br>

Answer:<br>
R_2 score for training data: 0.56 <br>
Intercept for linear regression: 143.51

---------------------------------------

In [75]:
y_test_pred = model.predict(X_test_scaled)
mae_test = mean_absolute_error(y_test, y_test_pred)
mean_demand = y_test.mean()
mae_test / mean_demand
r2_score(y_test, y_test_pred)

0.12697447171478002

0.5657413774267228

-------------------------------------
3) f) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer:<br>
Mean absolute error ratio for test data: 0.13
R_2 score for test data: 0.57

---------------------------------------

In [76]:
lasso_model = Lasso(alpha=0.2)
lasso_model.fit(X_train_scaled, y_train)
y_train_pred_lasso = lasso_model.predict(X_train_scaled)
r2_score(y_train, y_train_pred_lasso)
lasso_model.intercept_

Lasso(alpha=0.2)

0.5560135975109399

143.51450777202072

-------------------------------------
3) g) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer: <br>
Score for Lasso regression on train data: 0.56
Intercept for Lasso regression: 143.51

---------------------------------------

In [77]:
lasso_coefficients = lasso_model.coef_
feature_names = X.columns
[feature_names[i] for i in range(len(feature_names)) if lasso_coefficients[i] == 0]

['logLikes']

-------------------------------------
3) h) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer: <br>
Feature that got eliminated: 'logLikes'

---------------------------------------

In [78]:
y_test_pred_lasso = lasso_model.predict(X_test_scaled)
mae_test_lasso = mean_absolute_error(y_test, y_test_pred_lasso)
mae_test_lasso / mean_demand

0.1269708603261075

-------------------------------------
3) i) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer: <br>
Ratio of mean absolute error versus the mean of the Demand: 0.13

---------------------------------------

In [79]:
midterm_data = pd.read_excel('Kaggle_Data/demand_pred_midterm.xlsx')
midterm_data.dropna(inplace=True)
X_original = midterm_data.drop('Demand', axis=1) 
y_original = midterm_data['Demand']
categorical_cols = X_original.select_dtypes(include=['object']).columns
if len(categorical_cols) > 0:
    X_original.drop(columns=categorical_cols, inplace=True)
model = LinearRegression()
model.fit(X_original, y_original)
midterm_pred = pd.read_excel('Kaggle_Data/demand_pred_students.xlsx')
midterm_pred.dropna(inplace=True)
X_pred = midterm_pred 
categorical_cols_pred = X_pred.select_dtypes(include=['object']).columns
if len(categorical_cols_pred) > 0:
    X_pred.drop(columns=categorical_cols_pred, inplace=True)
model.predict(X_pred)


LinearRegression()

array([159.330701  , 105.37110875, 156.84255947, 116.87304638])

-------------------------------------
3) j) <br>
Name: Venkata Gade <br>
Email: vngade@syr.edu <br>
SUID: 979072739 

Answer: <br>
Four values of best guess: 159.330701, 105.37110875, 156.84255947, 116.87304638

---------------------------------------